### PR TITLE
Support for idle notify and idle inhibit

### PR DESCRIPTION
**Pull request by snaggen**
Initial support for idle notify and idle inhibit.

### CODE REVIEW

File: src/wayland/handlers/mod.rs [hunk 1]
```diff
@@ -9,6 +9,7 @@ pub mod dmabuf;
 pub mod drm;
 pub mod drm_lease;
 pub mod fractional_scale;
+pub mod idle;
```
Comment (Drakulix) [inline src/wayland/handlers/mod.rs:12]: Can you split this into two modules for `idle_notify` and `idle_inhibit` please?
I know they are related in how this implementation currently works, but they don't have to be.

**Review by Drakulix — COMMENTED** [review]
Looks mostly good! Thanks for working on this :)

**Comment by snaggen** [comment]
Inhibit is broken... seems the surface_primary_scanout_output function is not hooked up. And since I shamelessly stole this from niri, I must investigate this.

**Comment by snaggen** [comment]
Ok, By thinking about this over night, I got to the conclusion that niri is doing this wrong. There are many use case where you want to inhibit idle without being in full screen. So, I will modify this to allow any surface that is alive to inhibit idle.

**Comment by Drakulix** [comment]
> Tested with swayidle and idle seems to work as expected. Also, inihibit seems to work. I first thought it didn't but it seems the scanout check in the refresh only works for fullscreen.... so inhibit only works for fullscreen. Not sure if that is desired? I shamelessly stole that part of the code from niri, so I'm not sure what to think about it.

Hmm, it definitely shouldn't only work for fullscreen. The check basically makes sure the surface is actually visible, otherwise any client could just create a wl_surface and never map it any screen, which is undesirable.

So imo the check was fine, but it should only work for windows on the active workspace (or sticky windows). How were you testing?

**Comment by snaggen** [comment]
I was using https://github.com/stwa/wayland-idle-inhibitor since the video player i tested doesn't seem to inhibit. But, I'll add back the check then and see if it was my testing that was flawed.

**Comment by snaggen** [comment]
Ok, seems that it was flawed testing. The current MR works as expected when tested with Clapper, which uses GTK4 builtin Video playout. That seems to have proper support for the wayland inhibit protocol. I notice that firefox is supposed to have inhibit support, but I couldn't get that to work. But since clapper works properly, I assume the firefox issue is a client issue.

**Comment by snaggen** [comment]
About the failure with firefox, maybe @stransky can cast some light on this. Does firefox require some specific conditions besides just playing video, to inhibit idle?

**Comment by snaggen** [comment]
Also, a simple way to test this is to use swayidle like:
```bash
swayidle -w timeout 5 "echo idle" resume "echo resume"
```

**Comment by stransky** [comment]
> About the failure with firefox, maybe @stransky can cast some light on this. Does firefox require some specific conditions besides just playing video, to inhibit idle?

You can run Firefox on terminal with MOZ_LOG="LinuxWakeLock:5" env variable and you'll get info when Firefox sends inhibit/uninhibit. AFAIK the inhibit event is generated by DOM HTML video element and depends on visibility and so on, it may request more source investigation but it's possible to find it.

**Comment by YaLTeR** [comment]
hey, just a few drive by comments

> I got to the conclusion that niri is doing this wrong. There are many use case where you want to inhibit idle without being in full screen

Not really related to this PR, but in niri it works whenever the surface is anywhere on screen, not just fullscreen.

> firefox

When I was implementing idle, I found that Firefox always tries the portal inhibit first, and the portal inhibit literally never reports a failure (there are various bugs reported about this into the portals), so Firefox never got to the Wayland inhibit. I ended up also implementing the FDO D-Bus inhibit endpoint.

**Comment by snaggen** [comment]
Yes, my niri comment was wrong. I had misunderstood what it did. So, I went back to the same logic in this PR

**Comment by snaggen** [comment]
> When I was implementing idle, I found that Firefox always tries the portal inhibit first, and the portal inhibit literally never reports a failure (there are various bugs reported about this into the portals), so Firefox never got to the Wayland inhibit. I ended up also implementing the FDO D-Bus inhibit endpoint.

I'll see of I get some time to test Firefox a bit more. But @stransky , would it be possible to prefer Wayland inhibit, and fallback to dbus? Would that be more reliable?

**Comment by stransky** [comment]
> > When I was implementing idle, I found that Firefox always tries the portal inhibit first, and the portal inhibit literally never reports a failure (there are various bugs reported about this into the portals), so Firefox never got to the Wayland inhibit. I ended up also implementing the FDO D-Bus inhibit endpoint.
> 
> I'll see if I get some time to test Firefox a bit more. But @stransky , would it be possible to prefer Wayland inhibit, and fallback to dbus? Would that be more reliable?

DBus is what most of the apps use and it's preferred. It may be possible to add config to force-select particular inhibit type - please file a bug for it at https://bugzilla.mozilla.org/ and cc me there.

**Comment by Drakulix** [comment]
I think the proper solution for the firefox prooblem is to implement the dbus interface in xdg-desktop-portal-cosmic (utilizing this protocol), I don't think firefox needs to be adjusted here.

@snaggen Anything else you want to add to this PR? Or is this ready for review and potentially merge?

**Comment by snaggen** [comment]
This PR is ready for review.

**Review by Drakulix — APPROVED** [review]
Thanks for working on this!

**Pull request merged**